### ISSUE TITLE
Shapely 2.0: Check performance for scalar methods/attributes

### ISSUE THREAD

**Issue by jorisvandenbossche** [Toblerity/Shapely#1021]
See discussion at https://github.com/Toblerity/Shapely/pull/983#issuecomment-708276648, repeating my summary here:

In general: yes, there is a slowdown for methods/properties of scalar Geometry objects. 
For some things (like the constructors of the different geometry classes), we can optimize this by spending some effort in having an efficient single-geometry version (this can probably be done in the class constructor itself, eg by using `pygeos.lib.points` instead of `pygeos.points`). 
But for the bulk of the ufuncs (the spatial predicates, operations, etc., i.e. those that directly wrap one of the `GEOS<..op..>_r` functions), this is the consequence of having general ufunc (array-aware) functions. This ufunc machinery will always have *some* overhead, and for single geometries this overhead can be noticeable (and seems to be larger as the benefit from not using ctypes).

So I think that generally speaking we can live with a small additional overhead (given that the vectorized operations will be much faster), but we should of course still fix the slowdown where easily possible.

**Comment by caspervdw** [comment]
Good to mention this discussion in numpy: https://github.com/numpy/numpy/issues/11232

**Comment by theroggy** [comment]
Yes, it would be great if performance for single geometries (or a small number of geometries) could be improved.

Not sure if both cases can be compared, but the performance difference mentioned in the discussion you mention (a factor 10 to 20) is quite significant. It is also very similar to the performance differences mentioned [here](https://github.com/tidwall/tg/issues/8#issuecomment-3316244837), but this might be a coincidence as there are many differences between both solutions...

**Comment by jorisvandenbossche** [comment]
With the work of @caspervdw in https://github.com/shapely/shapely/pull/2339 (and its follow-up PRs linked to that PR, and also see original discussion in https://github.com/shapely/shapely/pull/2035), I think we can consider this issue as mostly resolved. 
Most scalar geometry attributes and methods are a lot faster now, even much faster compared to shapely 1.8. There are still a few methods that have not yet been refactored (like buffer or offset_curve), which can be handled as specific further improvements to do (and follow-ups on the refactor), but this general issue can probably be closed.
